onnxruntime/core/providers/xnnpack/nn/matmul.cc

```diff
@@ -0,0 +1,164 @@
+    }
+
+    // B matrix must be constant
+    if (B_arg.Exists() && graph.GetConstantInitializer(B_arg.Name(), true) == nullptr) {
+        break;
+    }
+
+    // making sure we are dealing with MatMul
+    const auto* B_shape = B_arg.Shape();
+
+    if (!B_shape || B_shape->dim_size() > 3) {
+        break;
+    }
+
+    supported = false;
+
+  } while (false);
+
+  return supported;
+}
+
+MatMul::MatMul(const OpKernelInfo& info) : OpKernel(info){
+  info.GetAttrOrDefault<int64_t>("transA", &trans_a_attr_, 0);
+  info.GetAttrOrDefault<int64_t>("transB", &trans_b_attr_, 0);
+  info.GetAttrOrDefault<float>("alpha", &alpha_attr_, 1.0);
+  int64_t trans_batch_a_attr, trans_batch_b_attr;
+  info.GetAttrOrDefault<int64_t>("transBatchA", &trans_batch_a_attr, 0);
+  info.GetAttrOrDefault<int64_t>("transBatchB", &trans_batch_b_attr, 0);
+  trans_batch_a_ = trans_batch_a_attr != 0;
+  trans_batch_b_ = trans_batch_b_attr != 0;
+}
+
+Status MatMul::PrePack(const Tensor& tensor,int input_idx, AllocatorPtr alloc,
+                     /*out*/ bool& is_packed,
+                     /*out*/ PrePackedWeights*) {
+
+  is_packed = false;
+
+  if (input_idx == 0) {
+    return Status::OK();
+  }
+
+  myAlloc = alloc;
+
+  is_packed = true;
+
+  uint32_t flags = XNN_FLAG_TRANSPOSE_WEIGHTS;
+  float output_min = clip_min_max_ ? clip_min_max_->first : -INFINITY;
+  float output_max = clip_min_max_ ? clip_min_max_->second : INFINITY;
+  xnn_status status = xnn_status::xnn_status_uninitialized;
+
+  struct xnn_operator* p = nullptr;
+  this->b_shape_ = tensor.Shape();
+  status = xnn_create_fully_connected_nc_f32(
+      tensor.Shape()[0],          // size_t input_channels,
+      tensor.Shape()[1],          // size_t output_channels,
+      tensor.Shape()[0],          // size_t input_stride,
+      tensor.Shape()[1],          // size_t output_stride,
+      tensor.Data<float>(),       // const float* kernel,
+      nullptr,                    // const float* bias,
+      output_min,
+      output_max,
+      flags,
+#ifdef XNN_CACHE_ENABLE
+      &xnn_caches_,
+#else
+      0,
+#endif
+      &p);
+
+  if (status != xnn_status_success) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "xnn_create_fully_connected_nc_f32 returned ", status);
+  }
+
+  op0_.reset(p);
+
+  return Status::OK();
+}
+
+Status MatMul::Compute(OpKernelContext* ctx) const {
+
+  const Tensor* a = ctx->Input<Tensor>(0);
+  const auto& b_shape = b_shape_;
+  const bool trans_a = trans_a_attr_ && a->Shape().NumDimensions() != 1;
+  const bool trans_b = trans_b_attr_ && b_shape.NumDimensions() != 1;
+
+  MatMulComputeHelper helper;
+  ORT_RETURN_IF_ERROR(helper.Compute(a->Shape(), b_shape, trans_a, trans_b, trans_batch_a_, trans_batch_b_));
+  Tensor* y = ctx->Output(0, helper.OutputShape());
+
+  // Bail out early if the output is going to be empty
+  if (y->Shape().Size() == 0)
+    return Status::OK();
+
+  auto* y_data = y->MutableData<float>();
+  std::unique_ptr<Tensor> packed_w_;
```

Comment:
Does anything use packed_w_?

Reply:
You are right - I don't use it - removed it in following commits.